kupyna/src/long.rs

```diff
@@ -1,95 +1,74 @@
-use crate::utils::{add_constant_plus, add_constant_xor, apply_s_box, mix_columns, xor_bytes};
+use crate::utils::{
+    add_constant_plus, add_constant_xor, apply_s_box, convert_message_block, mix_columns, xor,
+};
 
 pub(crate) const COLS: usize = 16;
 const ROUNDS: u64 = 14;
 
-type Matrix = [[u8; 8]; 16];
-
 pub(crate) fn compress(prev_vector: &mut [u64; COLS], message_block: &[u8; 128]) {
-    let mut prev_vector_u8 = [0u8; 128];
-    for (src, dst) in prev_vector.iter().zip(prev_vector_u8.chunks_exact_mut(8)) {
-        dst.copy_from_slice(&src.to_be_bytes());
-    }
+    // Convert message block from u8 to u64 (column-major order as per paper)
+    let message_u64 = convert_message_block::<128, COLS>(message_block);
 
-    let m_xor_p = xor_bytes(*message_block, prev_vector_u8);
+    let m_xor_p = xor(*prev_vector, message_u64);
 
     let t_xor_mp = t_xor_l(m_xor_p);
 
-    let t_plus_m = t_plus_l(*message_block);
+    let t_plus_m = t_plus_l(message_u64);
 
-    prev_vector_u8 = xor_bytes(xor_bytes(t_xor_mp, t_plus_m), prev_vector_u8);
-
-    for (dst, src) in prev_vector.iter_mut().zip(prev_vector_u8.chunks_exact(8)) {
-        *dst = u64::from_be_bytes(src.try_into().unwrap());
-    }
+    *prev_vector = xor(xor(t_xor_mp, t_plus_m), *prev_vector);
 }
 
-pub(crate) fn t_plus_l(block: [u8; 128]) -> [u8; 128] {
-    let mut state = block_to_matrix(block);
+pub(crate) fn t_plus_l(state: [u64; COLS]) -> [u64; COLS] {
+    let mut state = state;
     for nu in 0..ROUNDS {
         state = add_constant_plus(state, nu as usize);
         state = apply_s_box(state);
         state = rotate_rows(state);
         state = mix_columns(state);
     }
-    matrix_to_block(state)
+    state
 }
 
-fn block_to_matrix(block: [u8; 128]) -> Matrix {
-    const ROWS: usize = 16;
-    const COLS: usize = 8;
-
-    let mut matrix = [[0u8; COLS]; ROWS];
-    for i in 0..ROWS {
-        for j in 0..COLS {
-            matrix[i][j] = block[i * COLS + j];
+fn rotate_rows(state: [u64; COLS]) -> [u64; COLS] {
+    // Convert to matrix format (column-major as per paper)
+    let mut matrix = [[0u8; COLS]; 8];
+    for col in 0..COLS {
```

Comment:
If you want, you can use `#[allow(clippy::needless_range_loop)]` on this ans similar functions.